File: trading-algo/src/team/executor/paper.ts
import type { Order, Position, Portfolio, Signal, RiskAssessment } from '../../shared/types.js';
import type { ExecutorConfig, TradeExecution } from './types.js';
import { createOrderFromSignal, fillOrder } from './order.js';
import { generateId, roundTo } from '../../shared/utils.js';
import { createModuleLogger } from '../../shared/logger.js';
import { eventBus } from '../../shared/events.js';

const log = createModuleLogger('paper-trader');

/**
 * Paper trading engine — simulates real trading without risking capital.
 */
export class PaperTrader {
  private portfolio: Portfolio;
  private orderHistory: Order[] = [];
  private config: ExecutorConfig;
  private peakEquity: number;
  /** Highest price seen for each open position (for trailing stops). */
  private highWatermarks = new Map<string, number>();
  /** Initial ATR stop distance at entry price (entry - stopLoss). */
  private initialStopDistances = new Map<string, number>();

  constructor(config?: Partial<ExecutorConfig>) {
    this.config = {
      mode: 'paper',
      initialCapital: config?.initialCapital ?? 10000,
      maxOpenPositions: config?.maxOpenPositions ?? 10,
      defaultSlippage: config?.defaultSlippage ?? 0.0005,
      defaultCommission: config?.defaultCommission ?? 0.001,
    };

    this.peakEquity = this.config.initialCapital;

    this.portfolio = {
      capital: this.config.initialCapital,
      availableCapital: this.config.initialCapital,
      positions: [],
      totalPnl: 0,
      totalPnlPct: 0,
      maxDrawdown: 0,
      lastUpdated: Date.now(),
    };

    log.info({ capital: this.config.initialCapital }, 'Paper trader initialized');
  }

  /**
   * Execute a trade based on signal and risk assessment.
   */
  async executeTrade(signal: Signal, risk: RiskAssessment): Promise<TradeExecution> {
    if (!risk.approved) {
      log.warn({ reason: risk.reason }, 'Trade rejected by risk manager');
      return { order: {} as Order, portfolio: this.portfolio, success: false, error: risk.reason };
    }

    if (signal.action === 'HOLD') {
      return { order: {} as Order, portfolio: this.portfolio, success: false, error: 'HOLD signal' };
    }

    // Check max positions — try evicting a losing position before hard-rejecting
    if (signal.action === 'BUY' && this.portfolio.positions.length >= this.config.maxOpenPositions) {
      if (!this.evictWorstLoser()) {
        return { order: {} as Order, portfolio: this.portfolio, success: false, error: 'Max positions reached — no losers to evict' };
      }
    }

    const quantity = risk.recommendedSize / signal.price;
    const order = createOrderFromSignal(signal, quantity);

    // Simulate fill with slippage
    const slippage = signal.action === 'BUY' ? 1 + this.config.defaultSlippage : 1 - this.config.defaultSlippage;
    const fillPrice = signal.price * slippage;
    const commission = fillPrice * quantity * this.config.defaultCommission;

    if (signal.action === 'BUY') {
      const totalCost = fillPrice * quantity + commission;
      if (totalCost > this.portfolio.availableCapital) {
        return { order, portfolio: this.portfolio, success: false, error: 'Insufficient capital' };
      }

      const filledOrder = fillOrder(order, fillPrice, quantity);
      this.orderHistory.push(filledOrder);

      const position: Position = {
        id: generateId(),
        asset: signal.asset,
        side: 'buy',
        entryPrice: fillPrice,
        currentPrice: fillPrice,
        quantity,
        stopLoss: risk.stopLossPrice,
        takeProfit: risk.takeProfitPrice,
        unrealizedPnl: -commission,
        realizedPnl: 0,
        status: 'open',
        openedAt: Date.now(),
        strategy: signal.strategy,
      };

      this.portfolio.positions.push(position);
      this.portfolio.availableCapital -= totalCost;
      this.portfolio.lastUpdated = Date.now();

      // Seed trailing stop tracking
      this.highWatermarks.set(position.id, fillPrice);
      const stopDist = risk.stopLossPrice > 0 ? fillPrice - risk.stopLossPrice : 0;
      if (stopDist > 0) this.initialStopDistances.set(position.id, stopDist);

      await eventBus.emit('order:filled', filledOrder, 'paper-trader');
      await eventBus.emit('position:opened', position, 'paper-trader');

      log.info(
        { symbol: signal.asset.symbol, side: 'buy', price: fillPrice, quantity, strategy: signal.strategy },
        'Paper trade executed'
      );
      console.log(
        `[TRADE] BUY  ${signal.asset.symbol.padEnd(12)} @$${fillPrice.toFixed(4)}` +
        `  qty: ${quantity.toFixed(6)}  strategy: ${signal.strategy}` +
        (risk.stopLossPrice ? `  SL: $${risk.stopLossPrice.toFixed(4)}` : '') +
        (risk.takeProfitPrice ? `  TP: $${risk.takeProfitPrice.toFixed(4)}` : '')
      );

      return { order: filledOrder, position, portfolio: this.portfolio, success: true };
    } else {
      // SELL: close existing position
      const posIndex = this.portfolio.positions.findIndex(
        (p) => p.asset.symbol === signal.asset.symbol && p.status === 'open'
      );

      if (posIndex === -1) {
        return { order, portfolio: this.portfolio, success: false, error: 'No position to close' };
      }

      const position = this.portfolio.positions[posIndex];
      const pnl = (fillPrice - position.entryPrice) * position.quantity - commission;

      position.realizedPnl = pnl;
      position.status = 'closed';
      position.closedAt = Date.now();
      position.currentPrice = fillPrice;

      this.highWatermarks.delete(position.id);
      this.initialStopDistances.delete(position.id);

      this.portfolio.availableCapital += fillPrice * position.quantity - commission;
      this.portfolio.totalPnl += pnl;
      this.portfolio.totalPnlPct = (this.portfolio.totalPnl / this.config.initialCapital) * 100;
      this.portfolio.positions.splice(posIndex, 1);
      this.portfolio.lastUpdated = Date.now();

      const filledOrder = fillOrder(order, fillPrice, position.quantity);
      this.orderHistory.push(filledOrder);

      await eventBus.emit('order:filled', filledOrder, 'paper-trader');
      await eventBus.emit('position:closed', position, 'paper-trader');

      log.info(
        { symbol: signal.asset.symbol, side: 'sell', price: fillPrice, pnl: pnl.toFixed(2), strategy: signal.strategy },
        'Paper position closed'
      );
      const pnlSign = pnl >= 0 ? '+' : '';
      console.log(
        `[TRADE] SELL ${signal.asset.symbol.padEnd(12)} @$${fillPrice.toFixed(4)}` +
        `  PnL: ${pnlSign}$${pnl.toFixed(2)}  strategy: ${signal.strategy}`
      );

      return { order: filledOrder, position, portfolio: this.portfolio, success: true };
    }
  }

  /**
   * Update all position prices (called on market data update).
   */
  updatePrices(prices: Map<string, number>): void {
    for (const pos of this.portfolio.positions) {
      const price = prices.get(pos.asset.symbol);
      if (price !== undefined) {
        pos.currentPrice = price;
        pos.unrealizedPnl = (price - pos.entryPrice) * pos.quantity;

        // Ratchet trailing stop upward for long positions
        if (pos.side === 'buy' && pos.stopLoss !== undefined) {
          const hw = this.highWatermarks.get(pos.id) ?? pos.entryPrice;
          if (price > hw) {
            this.highWatermarks.set(pos.id, price);
            const stopDist = this.initialStopDistances.get(pos.id);
            if (stopDist) {
              const newStop = roundTo(price - stopDist, 6);
              if (newStop > pos.stopLoss) pos.stopLoss = newStop;
            }
          }
        }
      }
    }

    // Calculate total equity and drawdown
    const totalEquity = this.portfolio.availableCapital +
      this.portfolio.positions.reduce((sum, p) => sum + p.currentPrice * p.quantity, 0);

    this.portfolio.capital = totalEquity;
    if (totalEquity > this.peakEquity) this.peakEquity = totalEquity;
    const drawdown = ((this.peakEquity - totalEquity) / this.peakEquity) * 100;
    if (drawdown > this.portfolio.maxDrawdown) {
      this.portfolio.maxDrawdown = drawdown;
    }
    this.portfolio.lastUpdated = Date.now();
  }

  /**
   * Check and trigger stop losses / take profits.
   */
  async checkStops(currentPrices: Map<string, number>): Promise<void> {
    const toClose: Position[] = [];

    for (const pos of this.portfolio.positions) {
      const price = currentPrices.get(pos.asset.symbol);
      if (!price) continue;

      if (pos.stopLoss && price <= pos.stopLoss) {
        log.info({ symbol: pos.asset.symbol, price, stopLoss: pos.stopLoss }, 'Stop loss triggered');
        console.log(`[STOP-LOSS] ${pos.asset.symbol}  price: $${price.toFixed(4)}  SL: $${pos.stopLoss.toFixed(4)}`);
        toClose.push(pos);
      } else if (pos.takeProfit && price >= pos.takeProfit) {
        log.info({ symbol: pos.asset.symbol, price, takeProfit: pos.takeProfit }, 'Take profit triggered');
        console.log(`[TAKE-PROFIT] ${pos.asset.symbol}  price: $${price.toFixed(4)}  TP: $${pos.takeProfit.toFixed(4)}`);
        toClose.push(pos);
      }
    }

    for (const pos of toClose) {
      const price = currentPrices.get(pos.asset.symbol)!;
      const signal: Signal = {
        asset: pos.asset,
        action: 'SELL',
        confidence: 1,
        price,
        timestamp: Date.now(),
        strategy: pos.strategy,
        timeframe: '1h',
        indicators: {},
        reason: pos.stopLoss && price <= pos.stopLoss ? 'Stop loss triggered' : 'Take profit triggered',
      };
      await this.executeTrade(signal, {
        maxPositionSize: 0, recommendedSize: 0,
        stopLossPrice: 0, takeProfitPrice: 0,
        riskRewardRatio: 0, kellyFraction: 0,
        approved: true, reason: 'Closing position',
      });
    }
  }

  /**
   * Evict the worst losing position (>0.3% loss) to free capacity for a better signal.
   * Returns true if a position was closed, false if all positions are profitable.
   */
  private evictWorstLoser(): boolean {
    const losers = this.portfolio.positions.filter(
      (p) => p.unrealizedPnl < -(p.entryPrice * p.quantity * 0.003),
    );
    if (losers.length === 0) return false;

    const worst = losers.reduce((a, b) => (a.unrealizedPnl < b.unrealizedPnl ? a : b));
    const price = worst.currentPrice;
    const commission = price * worst.quantity * this.config.defaultCommission;
    const pnl = (price - worst.entryPrice) * worst.quantity - commission;

    worst.realizedPnl = pnl;
    worst.status = 'closed';
    worst.closedAt = Date.now();

    const idx = this.portfolio.positions.indexOf(worst);
    this.portfolio.positions.splice(idx, 1);
    this.portfolio.availableCapital += price * worst.quantity - commission;
    this.portfolio.totalPnl += pnl;
    this.portfolio.totalPnlPct = (this.portfolio.totalPnl / this.config.initialCapital) * 100;
    this.portfolio.lastUpdated = Date.now();

    this.highWatermarks.delete(worst.id);
    this.initialStopDistances.delete(worst.id);

    log.info(
      { symbol: worst.asset.symbol, pnl: pnl.toFixed(2), remaining: this.portfolio.positions.length },
      'Evicted losing position to free slot for new signal',
    );
    void eventBus.emit('position:closed', worst, 'paper-trader');
    return true;
  }

  getPortfolio(): Portfolio {
    return { ...this.portfolio };
  }

  getOrderHistory(): Order[] {
    return [...this.orderHistory];
  }

  getSummary(): string {
    const p = this.portfolio;
    return [
      `=== Paper Trading Summary ===`,
      `Capital: $${p.capital.toFixed(2)}`,
      `Available: $${p.availableCapital.toFixed(2)}`,
      `Open Positions: ${p.positions.length}`,
      `Total P&L: $${p.totalPnl.toFixed(2)} (${p.totalPnlPct.toFixed(2)}%)`,
      `Max Drawdown: ${p.maxDrawdown.toFixed(2)}%`,
      `Total Orders: ${this.orderHistory.length}`,
    ].join('\n');
  }
}

File: trading-algo-py/team/executor/paper.py
"""Paper trading engine — simulates order execution with margin tracking."""

from __future__ import annotations

import logging
import time
import uuid

from shared.types import (
    Signal, SignalAction, RiskAssessment, Order, Position, Portfolio,
    Side, OrderType, OrderStatus, PositionStatus,
)
from shared.events import event_bus
from team.risk_manager.margin import MarginManager, get_pair_leverage

log = logging.getLogger(__name__)


class PaperTrader:
    """Simulates trade execution with slippage, commission, and margin tracking."""

    def __init__(
        self,
        initial_capital: float = 10_000,
        slippage: float = 0.0005,
        commission: float = 0.001,
        max_open_positions: int = 10,
    ) -> None:
        self.initial_capital = initial_capital
        self.available_capital = initial_capital
        self.slippage = slippage
        self.commission = commission
        self.max_open = max_open_positions

        self.positions: list[Position] = []
        self.closed_positions: list[Position] = []
        self.orders: list[Order] = []
        self.peak_equity = initial_capital
        self.max_drawdown = 0.0
        self.margin_manager = MarginManager()

        log.info("PaperTrader initialised ($%.2f capital)", initial_capital)

    def execute_trade(self, signal: Signal, risk: RiskAssessment) -> dict:
        """Execute a paper trade based on signal and risk assessment."""
        if signal.action == SignalAction.HOLD:
            return {"success": False, "error": "HOLD signal"}

        if not risk.approved:
            return {"success": False, "error": risk.reason}

        open_count = sum(1 for p in self.positions if p.status == PositionStatus.OPEN)
        if open_count >= self.max_open:
            return {"success": False, "error": f"Max positions reached ({self.max_open})"}

        # Calculate position size
        quantity = risk.recommended_size / signal.price if signal.price > 0 else 0
        if quantity <= 0:
            return {"success": False, "error": "Position size too small"}

        # Apply slippage
        if signal.action == SignalAction.BUY:
            fill_price = signal.price * (1 + self.slippage)
            side = Side.BUY
        else:
            fill_price = signal.price * (1 - self.slippage)
            side = Side.SELL

        # Leverage & margin calculation
        leverage = get_pair_leverage(signal.asset)
        notional = quantity * fill_price
        margin_req = notional / leverage

        # Commission
        commission_cost = notional * self.commission
        total_cost = margin_req + commission_cost  # Only margin + commission locked (not full notional)

        # Margin check
        portfolio = self.get_portfolio()
        margin_check = self.margin_manager.check_margin(
            signal.asset, quantity, fill_price, portfolio,
        )
        if not margin_check.allowed:
            return {"success": False, "error": f"Margin: {margin_check.reason}"}

        if total_cost > self.available_capital:
            return {"success": False, "error": f"Insufficient free margin: need ${total_cost:.2f}, have ${self.available_capital:.2f}"}

        # Create order
        order = Order(
            id=str(uuid.uuid4())[:8],
            asset=signal.asset,
            side=side,
            type=OrderType.MARKET,
            quantity=quantity,
            strategy=signal.strategy,
            price=signal.price,
            status=OrderStatus.FILLED,
            filled_price=fill_price,
            filled_quantity=quantity,
            created_at=int(time.time() * 1000),
            filled_at=int(time.time() * 1000),
        )
        self.orders.append(order)

        # Create position with margin tracking
        position = Position(
            id=str(uuid.uuid4())[:8],
            asset=signal.asset,
            side=side,
            entry_price=fill_price,
            current_price=fill_price,
            quantity=quantity,
            strategy=signal.strategy,
            stop_loss=risk.stop_loss_price,
            take_profit=risk.take_profit_price,
            opened_at=int(time.time() * 1000),
            leverage=leverage,
            margin_required=margin_req,
            notional_value=notional,
        )
        self.positions.append(position)
        self.available_capital -= total_cost

        event_bus.emit("order:filled", order, "PaperTrader")
        event_bus.emit("position:opened", position, "PaperTrader")

        log.info(
            "%s %s %.4f @ %.5f (SL=%.5f, TP=%.5f)",
            side.value.upper(), signal.asset.symbol, quantity, fill_price,
            risk.stop_loss_price, risk.take_profit_price,
        )

        return {"success": True, "order": order, "position": position, "portfolio": self.get_portfolio()}

    def update_prices(self, prices: dict[str, float]) -> None:
        """Update current prices for all open positions."""
        for pos in self.positions:
            if pos.status == PositionStatus.OPEN and pos.asset.symbol in prices:
                pos.current_price = prices[pos.asset.symbol]
                if pos.side == Side.BUY:
                    pos.unrealized_pnl = (pos.current_price - pos.entry_price) * pos.quantity
                else:
                    pos.unrealized_pnl = (pos.entry_price - pos.current_price) * pos.quantity

    def check_stops(self, prices: dict[str, float]) -> list[Position]:
        """Check stop loss and take profit levels."""
        self.update_prices(prices)
        closed: list[Position] = []

        for pos in self.positions:
            if pos.status != PositionStatus.OPEN:
                continue

            price = pos.current_price
            hit_sl = pos.stop_loss and (
                (pos.side == Side.BUY and price <= pos.stop_loss) or
                (pos.side == Side.SELL and price >= pos.stop_loss)
            )
            hit_tp = pos.take_profit and (
                (pos.side == Side.BUY and price >= pos.take_profit) or
                (pos.side == Side.SELL and price <= pos.take_profit)
            )

            if hit_sl or hit_tp:
                self._close_position(pos, price, "stop_loss" if hit_sl else "take_profit")
                closed.append(pos)

        return closed

    def _close_position(self, pos: Position, price: float, reason: str) -> None:
        """Close a position at the given price."""
        pos.current_price = price
        if pos.side == Side.BUY:
            pos.realized_pnl = (price - pos.entry_price) * pos.quantity
        else:
            pos.realized_pnl = (pos.entry_price - price) * pos.quantity

        commission = price * pos.quantity * self.commission
        pos.realized_pnl -= commission

        pos.unrealized_pnl = 0
        pos.status = PositionStatus.CLOSED
        pos.closed_at = int(time.time() * 1000)

        # Release margin + return PnL (margin was locked, not full notional)
        self.available_capital += pos.margin_required + pos.realized_pnl
        self.closed_positions.append(pos)

        event_bus.emit("position:closed", pos, "PaperTrader")
        log.info("Closed %s %s @ %.5f (%s, PnL=%.2f)",
                 pos.side.value, pos.asset.symbol, price, reason, pos.realized_pnl)

    def get_portfolio(self) -> Portfolio:
        """Get current portfolio state with margin tracking."""
        open_positions = [p for p in self.positions if p.status == PositionStatus.OPEN]

        # Update notional values for open positions
        for p in open_positions:
            p.notional_value = p.current_price * p.quantity

        margin_used = sum(p.margin_required for p in open_positions)
        total_notional = sum(p.notional_value for p in open_positions)
        unrealized = sum(p.unrealized_pnl for p in open_positions)
        realized = sum(p.realized_pnl for p in self.closed_positions)
        total_pnl = unrealized + realized

        # Equity = available cash + margin locked + unrealized PnL
        equity = self.available_capital + margin_used + unrealized
        margin_available = max(0, equity - margin_used)
        margin_level = (equity / margin_used * 100) if margin_used > 0 else 0.0
        effective_leverage = total_notional / equity if equity > 0 else 0.0

        # Track drawdown
        if equity > self.peak_equity:
            self.peak_equity = equity
        drawdown = self.peak_equity - equity
        if drawdown > self.max_drawdown:
            self.max_drawdown = drawdown

        portfolio = Portfolio(
            capital=equity,
            available_capital=self.available_capital,
            positions=open_positions,
            total_pnl=total_pnl,
            total_pnl_pct=(total_pnl / self.initial_capital) * 100,
            max_drawdown=self.max_drawdown,
            last_updated=int(time.time() * 1000),
            margin_used=margin_used,
            margin_available=margin_available,
            margin_level_pct=margin_level,
            total_leverage=effective_leverage,
            total_notional=total_notional,
        )

        # Check stop-out
        stop_out = self.margin_manager.get_stop_out_positions(portfolio)
        if stop_out:
            for pos in stop_out[:1]:  # Close worst position
                self._close_position(pos, pos.current_price, "margin_stop_out")
                log.warning("MARGIN STOP-OUT: force-closed %s %s", pos.side.value, pos.asset.symbol)
            return self.get_portfolio()  # Recalculate after close

        return portfolio

    def get_order_history(self) -> list[Order]:
        return list(self.orders)

    def get_summary(self) -> str:
        p = self.get_portfolio()
        wins = sum(1 for pos in self.closed_positions if pos.realized_pnl > 0)
        losses = sum(1 for pos in self.closed_positions if pos.realized_pnl <= 0)
        total = wins + losses
        win_rate = (wins / total * 100) if total > 0 else 0
        margin_str = self.margin_manager.get_margin_summary(p)
        return (
            f"Portfolio: ${p.capital:.2f} | PnL: ${p.total_pnl:.2f} ({p.total_pnl_pct:.1f}%) | "
            f"Open: {len(p.positions)} | Closed: {total} | WR: {win_rate:.0f}% | DD: ${p.max_drawdown:.2f}\n"
            f"  {margin_str}"
        )
